File: bot/middleware.py
import logging
import time
from collections import defaultdict

from telegram import Update
from telegram.ext import ContextTypes

logger = logging.getLogger(__name__)

# Simple in-memory rate limiter: user_id -> list of timestamps
_rate_limit_windows: dict[int, list[float]] = defaultdict(list)
RATE_LIMIT_MAX = 30       # max messages
RATE_LIMIT_WINDOW = 60.0  # per 60 seconds


async def rate_limit_check(update: Update, context: ContextTypes.DEFAULT_TYPE) -> bool:
    """Return True if the message should be processed, False if rate-limited."""
    user_id = update.effective_user.id
    now = time.time()

    # Clean old timestamps
    _rate_limit_windows[user_id] = [
        t for t in _rate_limit_windows[user_id] if now - t < RATE_LIMIT_WINDOW
    ]

    if len(_rate_limit_windows[user_id]) >= RATE_LIMIT_MAX:
        logger.warning("Rate limited user %d", user_id)
        return False

    _rate_limit_windows[user_id].append(now)
    return True


async def log_incoming(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Log every incoming message for debugging."""
    user = update.effective_user
    text = update.message.text or update.message.caption or "[non-text]"
    # Truncate long messages in logs
    display = text[:200] + "..." if len(text) > 200 else text
    logger.info(
        "IN  | user=%d (@%s) | %s",
        user.id,
        user.username or "N/A",
        display,
    )


async def log_outgoing(user_id: int, text: str):
    """Log every outgoing message for debugging."""
    display = text[:200] + "..." if len(text) > 200 else text
    logger.info("OUT | user=%d | %s", user_id, display)

File: src/bot/middleware.ts
import { Context } from "telegraf";
import { Logger } from "../utils/logger";

const rateLimitWindows = new Map<number, number[]>();
const RATE_LIMIT_MAX = 30;       // max messages
const RATE_LIMIT_WINDOW = 60000; // per 60 seconds (in ms)

export async function rateLimitCheck(ctx: Context): Promise<boolean> {
    if (!ctx.from) return true;
    
    const userId = ctx.from.id;
    const now = Date.now();

    let timestamps = rateLimitWindows.get(userId) || [];
    
    // Clean old timestamps
    timestamps = timestamps.filter(t => now - t < RATE_LIMIT_WINDOW);
    
    if (timestamps.length >= RATE_LIMIT_MAX) {
        Logger.warn(`Rate limited user ${userId}`);
        rateLimitWindows.set(userId, timestamps);
        return false;
    }

    timestamps.push(now);
    rateLimitWindows.set(userId, timestamps);
    return true;
}

export async function logIncoming(ctx: Context): Promise<void> {
    if (!ctx.from || !ctx.message) return;
    
    const user = ctx.from;
    const text = 'text' in ctx.message ? ctx.message.text : 
                 'caption' in ctx.message ? ctx.message.caption : "[non-text]";
    
    const display = text && text.length > 200 ? text.substring(0, 200) + "..." : text || "[non-text]";
    
    Logger.info(`IN  | user=${user.id} (@${user.username || "N/A"}) | ${display}`);
}

export async function logOutgoing(userId: number, text: string): Promise<void> {
    const display = text.length > 200 ? text.substring(0, 200) + "..." : text;
    Logger.info(`OUT | user=${userId} | ${display}`);
}
